### DISCUSSION TITLE
[Run issue]: -bash: topas: command not found

### DISCUSSION THREAD

**Discussion by jonahpeter** [OpenTOPAS/OpenTOPAS#35] [Run-Problem] [answered]
### Contact Details

jonahpeter@g.harvard.edu

### What was the issue?

Hi all, I attempted to build OpenTOPAS for MacOS following the QuickStart Guide. I was able to complete all steps without any errors but when I try to run an example I get the following error:

`-bash: topas: command not found`

I then tried defining the environment variables manually as in Step 8.1 but when I run the bin/topas executable I get

`dyld[12395]: Library not loaded: @rpath/libQt5OpenGL.5.dylib
  Referenced from: <273764A8-3444-302B-B61B-42B76C8E73DC> /Applications/TOPAS/OpenTOPAS-install/bin/topas
  Reason: tried: '/Applications/GEANT4/geant4-install/lib/libQt5OpenGL.5.dylib' (no such file), '/Applications/TOPAS/OpenTOPAS-install/lib/libQt5OpenGL.5.dylib' (no such file), '/libQt5OpenGL.5.dylib' (no such file), '/Applications/TOPAS/OpenTOPAS-install/Frameworks/libQt5OpenGL.5.dylib' (no such file), '/Applications/TOPAS/OpenTOPAS-install/Frameworks/libQt5OpenGL.5.dylib' (no such file), '/usr/local/lib/libQt5OpenGL.5.dylib' (no such file), '/usr/lib/libQt5OpenGL.5.dylib' (no such file, not in dyld cache)
Abort trap: 6`

Anyone know what's going on?

### TOPAS category

Geometry

### Version

4.0

### What Operative System (OS) are you running on?

MacOS

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Attestation

- [X] I testify that I have fully read the bug reporting guideline beforehand.

**Comment by JoseRamosMendez** [comment]
Hi there, do you have another version of Qt (6)? or did you install it from a source other than Homebrew?

**Comment by jonahpeter** [comment]
Hi Jose, there is no output when I run `brew list --versions qt@6` and when I run `brew list --versions qt@5` I get the output: `qt@5 5.15.13_1`.

I used to have a version of Qt on my laptop but I thought I had uninstalled it properly beforehand... I tried running `brew uninstall qt@5` followed by `brew install qt@5` which yields the following error:

`Error: Not a directory @ dir_s_rmdir - /usr/local/Cellar/gst-libav`

So indeed it seems to be a Qt issue. Is there a way to do a full wipe of Qt on my system and reinstall via Homebrew?

Many thanks!
Jonah

**Comment by jonahpeter** [comment]
I tried uninstalling and reinstalling homebrew and encountered this error at step 6.2

`make[2]: *** No rule to make target `/usr/local/lib/libX11.dylib', needed by `BuildProducts/lib/libG4RayTracer.dylib'.  Stop.
make[1]: *** [source/CMakeFiles/G4RayTracer.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target `/usr/local/lib/libX11.dylib', needed by `BuildProducts/lib/libG4ToolsSG.dylib'.  Stop.
make[1]: *** [source/CMakeFiles/G4ToolsSG.dir/all] Error 2
make[2]: *** No rule to make target `/usr/local/lib/libX11.dylib', needed by `BuildProducts/lib/libG4OpenGL.dylib'.  Stop.
make[1]: *** [source/CMakeFiles/G4OpenGL.dir/all] Error 2
[ 97%] Built target G4physicslists
make: *** [all] Error 2`

**Comment by jonahpeter** [comment, marked as answer]
Okay another update: I updated Xcode, uninstalled and reinstalled homebrew, deleted the Geant4, GDCM, and TOPAS build/install directories, and ran everything from scratch. I can now run the example file when defining the environment variables manually (woohoo!)

However, I do encounter the following topas error when running `topas Dose.txt` from the `/Applications/TOPAS/OpenTOPAS-install/examples/Scoring` directory:

`Welcome to TOPAS, Tool for Particle Simulation (Version 4.0)
Loading parameters starting from: TsUserParameters.txt
Topas quitting. Unable to find top level parameter file.
Specify the top level parameter file on the command line,
or provide a file named TsUserParameters.txt.

Welcome to TOPAS, Tool for Particle Simulation (Version 4.0)
Loading parameters starting from: TsUserParameters.txt
Topas quitting. Unable to find top level parameter file.
Specify the top level parameter file on the command line,
or provide a file named TsUserParameters.txt.`

**Reply by icmeyer** [reply]
Glad the code is compiling! Getting the depedencies configured correctly is not easy especially when you have previous installations lying around

I think 'TsUserParameters.txt' is the default parameter input name if you don't pass an input file name as an argument. Are you sure you are in the directory and typing out the full line `topas Dose.txt`? 

Here is an example of what running the file might look like:
<img width="1148" alt="image" src="https://github.com/OpenTOPAS/OpenTOPAS/assets/34250777/ecc60e2e-a225-4f9c-a22e-4dfc6f31d27f">

**Reply by jonahpeter** [reply]
Hi Isaac,

Many thanks for your reply. Here the output I get by running the same code as you:

```
(base) dhcp-10-250-246-187:Scoring jonahpeter$ which topas
/Users/jonahpeter/shellScripts/topas
(base) dhcp-10-250-246-187:Scoring jonahpeter$ pwd
/Applications/TOPAS/OpenTOPAS/examples/Scoring
(base) dhcp-10-250-246-187:Scoring jonahpeter$ ls Dose*
Dose.txt			DoseToMediumVsWater.txt
DoseInVoxelMaterials.txt	DoseVolumeHistogram.txt
(base) dhcp-10-250-246-187:Scoring jonahpeter$ topas Dose.txt

Welcome to TOPAS, Tool for Particle Simulation (Version 4.0)
Loading parameters starting from: TsUserParameters.txt
Topas quitting. Unable to find top level parameter file.
Specify the top level parameter file on the command line,
or provide a file named TsUserParameters.txt.

Welcome to TOPAS, Tool for Particle Simulation (Version 4.0)
Loading parameters starting from: TsUserParameters.txt
Topas quitting. Unable to find top level parameter file.
Specify the top level parameter file on the command line,
or provide a file named TsUserParameters.txt.
```

I get a similar error when running `cd /Applications/TOPAS-nBio/` followed by `source rundemos.sh`

**Reply by icmeyer** [reply]
Ah, must be something with the shell script. Would you mind copying and posting the contents of `/Users/jonahpeter/shellScripts/topas`?

**Reply by jonahpeter** [reply]
Yes of course, here it is:

```
(base) dhcp-10-250-246-187:water_radiolysis jonahpeter$ cat /Users/jonahpeter/shellScripts/topas
export QT_QPA_PLATFORM_PLUGIN_PATH=/Applications/TOPAS/OpenTOPAS-install/Frameworks
export TOPAS_G4_DATA_DIR=/Applications/GEANT4/G4DATA
export DYLD_LIBRARY_PATH=/Applications/TOPAS/OpenTOPAS-install/lib:
export DYLD_LIBRARY_PATH=/Applications/GEANT4/geant4-install/lib:
/Applications/TOPAS/OpenTOPAS-install/bin/topas
export QT_QPA_PLATFORM_PLUGIN_PATH=/Applications/TOPAS/OpenTOPAS-install/Frameworks
export TOPAS_G4_DATA_DIR=/Applications/GEANT4/G4DATA
export DYLD_LIBRARY_PATH=/Applications/TOPAS/OpenTOPAS-install/lib:/Applications/GEANT4/geant4-install/lib:/Applications/TOPAS/OpenTOPAS-install/lib:
export DYLD_LIBRARY_PATH=/Applications/GEANT4/geant4-install/lib:/Applications/GEANT4/geant4-install/lib:/Applications/TOPAS/OpenTOPAS-install/lib:
/Applications/TOPAS/OpenTOPAS-install/bin/topas
```

**Reply by icmeyer** [reply]
Thank you! Should be an easy fix, basically from the install instructions the last line will need a `$1` to pass the argument to the script to the topas executable. Looks like somehow your `LIBRARY_PATH` ended up different from the instructions as well. If you open the script with a text editor you will just need the following block:

```
export QT_QPA_PLATFORM_PLUGIN_PATH=/Applications/TOPAS/OpenTOPAS-install/Frameworks
export TOPAS_G4_DATA_DIR=/Applications/GEANT4/G4DATA
export DYLD_LIBRARY_PATH=/Applications/TOPAS/OpenTOPAS-install/lib:$DYLD_LIBRARY_PATH
export DYLD_LIBRARY_PATH=/Applications/GEANT4/geant4-install/lib:$DYLD_LIBRARY_PATH
/Applications/TOPAS/OpenTOPAS-install/bin/topas $1
```

**Reply by jonahpeter** [reply]
yes that worked, thank you!!

I do still get the following error when running the TOPAS-nBio demos:

```
(base) dhcp-10-250-246-187:Scoring jonahpeter$ cd /Applications/TOPAS-nBio/
(base) dhcp-10-250-246-187:TOPAS-nBio jonahpeter$ source rundemos.sh
Topas quitting, unable to open parameter file:../../scorers/DNAScorerStepByStep.txt
--failed -- dna/CharltonDNA.txt
Topas quitting, unable to open parameter file:../../scorers/DNAScorerStepByStep.txt
--failed -- dna/LinearDNA.txt
Topas quitting, unable to open parameter file:../../scorers/DNAScorerStepByStep.txt
--failed -- dna/CircularPlasmid.txt
^C
```

but maybe I should open a separate ticket for this

**Reply by JoseRamosMendez** [reply]
Yes, please open a new ticket in the TOPAS-nBio repository's discussions. 
Best!
